pkg/kubefed2/federate/util.go

```diff
@@ -49,3 +58,104 @@ func SetBasicMetaFields(resource *unstructured.Unstructured, apiResource metav1.
+	apiResourceLists, err := enable.GetServerPreferredResources(config)
+	if err != nil {
+		return nil, err
+	}
+
+	apiResources := make(map[string]metav1.APIResource)
+	for _, apiResourceList := range apiResourceLists {
+		if len(apiResourceList.APIResources) == 0 {
+			continue
+		}
+
+		gv, err := schema.ParseGroupVersion(apiResourceList.GroupVersion)
+		if err != nil {
+			return nil, errors.Wrap(err, "Error parsing GroupVersion")
+		}
+
+		group := gv.Group
+		for _, apiResource := range apiResourceList.APIResources {
+			if !apiResource.Namespaced || isAlreadyFederated(apiResource) ||
+				apiResourceMatchesSkipNames(apiResource, skipAPIResourceNames, group) {
+				continue
+			}
+
+			// TODO(irfanurrehman): Define a strategy to federate the latest available version
+			// Include a kind only once (if found under multiple group/versions).
+			if _, ok := apiResources[apiResource.Kind]; ok {
+				continue
+			}
+
+			// The individual apiResources do not have the group and version set
+			apiResource.Group = group
+			apiResource.Version = gv.Version
+
+			apiResources[apiResource.Kind] = apiResource
+		}
+	}
+
+	return apiResources, nil
+}
+
+func apiResourceMatchesSkipNames(apiResource metav1.APIResource, userGivenNames, group string) bool {
+	names := append(controllerCreatedAPIResourceNames, strings.Split(userGivenNames, ",")...)
+	for _, name := range names {
+		if name == "" {
+			continue
+		}
+		if enable.NameMatchesResource(name, apiResource, group) {
+			return true
+		}
+	}
+	return false
+}
+
+func isAlreadyFederated(apiResource metav1.APIResource) bool {
+	return strings.HasPrefix(apiResource.Kind, federationKindPrefix)
+}
+
+// resources stores a list of resources for an api type
+type resources struct {
+	// resource type information
+	apiResource metav1.APIResource
+	// resource list
+	resources []*unstructured.Unstructured
+}
+
+func getResourcesInNamespace(config *rest.Config, namespace string, skipAPIResourceNames string) ([]resources, error) {
+	apiResources, err := getAPIResourceList(config, skipAPIResourceNames)
+	if err != nil {
+		return nil, err
+	}
+
+	resourcesInNamespace := []resources{}
+	for _, apiResource := range apiResources {
+		client, err := ctlutil.NewResourceClient(config, &apiResource)
+		if err != nil {
+			return nil, errors.Wrapf(err, "Error creating client for %s", apiResource.Kind)
+		}
+
+		resourceList, err := client.Resources(namespace).List(metav1.ListOptions{})
+		if apierrors.IsNotFound(err) || resourceList == nil {
+			continue
+		}
+		if err != nil {
+			return nil, errors.Wrapf(err, "Error listing resources for %s", apiResource.Kind)
+		}
+
+		targetResources := resources{apiResource: apiResource}
+		for _, resource := range resourceList.Items {
+			targetResources.resources = append(targetResources.resources, &resource)
+		}
+
+		// It would be a waste of cycles to iterate through empty slices while federating resource
```

Comment:
(No action required) Consider checking the length of `resourceList.Items` after retrieval and continuing if 0.

Reply:
Updated! Thanks for the suggestion.